utils/array_utils.h

```diff
@@ -285,7 +306,8 @@ inline void flip(float *a, size_t R, size_t C, bool horizontal = true) {
  * @param tol The tolerance for closeness.
  * @return bool True if the arrays are close, false otherwise.
  */
-bool isclose(float *a, float *b, size_t n, float tol = 1e-3) {
+template<class precision=float>
```

Comment:
I think `isclose` can be kept as two concrete overloads (one for half, one for f32) - keeps this code easier to quickly scan and not get tripped up on having to think about which template overload is going to get triggered. If in the future we have more precision types, and this trival overloading starts is too repetitive, we can re-evaluate then.

But optimize for easy-to-scan and reason about is valuable enough to keep simple overloads the default, unless there's something i'm missing.